method_meta add uses_backend function

cmt0 opened the pull request:
Summary: For a runtime that supports models of a couple different types it helps to be able to check ahead of time whether certain backend types are present. In our case, there are processors that can use this information to make scheduling decisions without having the actual required backend implementations present.

Differential Revision: D69143825

dbort commented on runtime/executor/method_meta.h:
```diff
@@ -192,6 +192,13 @@ class MethodMeta final {
+   *
+   * @returns Whether a backend is used in this method.
+   */
+  ET_EXPERIMENTAL bool contains_backend(const char* backend_name) const;
```
Please call it `uses_backend`: a method doesn't "contain" a backend per se.

dbort commented on runtime/executor/method_meta.h:
```diff
@@ -192,6 +192,13 @@ class MethodMeta final {
    */
   ET_EXPERIMENTAL size_t num_instructions() const;
 
+  /**
+   * Check to see if a backend is used in this method.
+   *
+   * @returns Whether a backend is used in this method.
+   */
```
Please add a `@param`, and clarify true-vs-false rather than "whether".

```suggestion
  /**
   * Check to see if a backend is used by this method.
   *
   * @param backend_name The name of the backend to search for.
   * @returns true if this method uses this backend; false otherwise.
   */
```

cmt0 replied:
Added clarification

dbort replied:
Please add the @param line too

dbort commented on runtime/executor/method_meta.cpp:
```diff
@@ -248,5 +248,16 @@ size_t MethodMeta::num_instructions() const {
   return num_instructions;
 }
 
+bool MethodMeta::contains_backend(const char* backend_name) const {
```
Please add a test for this new method.

You could add a couple lines to https://github.com/cmt0/executorch/blob/09938978bf530202032943c89c5e7f482d1032a2/runtime/executor/test/backend_integration_test.cpp#L343, checking that it's true for a backend named `kName`, and that it's false for some other string like "INVALID_BACKEND_NAME".

cmt0 replied:
Good idea, added a test which does this

dbort reviewed (changes requested):
Thanks for adding this! It's a reasonable feature.

cmt0 commented:
@pytorchbot label "release notes: Adds method meta function which checks whether a backend is used by a program"

cmt0 commented:
@pytorchbot label "release notes: misc"

cmt0 commented:
@dbort Does it need an ET_EXPERIMENTAL tag, or can I define it normall to avoid deprecated declarations warning?

dbort commented:
> @dbort Does it need an ET_EXPERIMENTAL tag, or can I define it normall to avoid deprecated declarations warning?

I think the new method is general enough to remove the ET_EXPERIMENTAL annotation. The behavior is well-defined compared to e.g. the instruction count.

cmt0 committed:
method_meta add uses_backend function (#8198)

Summary:

For a runtime that supports models of a couple different types it helps to be able to check ahead of time whether certain backend types are present. In our case, there are processors that can use this information to make scheduling decisions without having the actual required backend implementations present.

Reviewed By: dbort

Differential Revision: D69143825
